common/network-common/src/main/java/org/apache/spark/network/sasl/SparkSaslAES.java

```diff
@@ -0,0 +1,243 @@
+    // append seqNum used for mac
+    byte[] wrapped = new byte[encrypted.length + 4];
+    System.arraycopy(encrypted, 0, wrapped, 0, encrypted.length);
+    System.arraycopy(integrity.getSeqNum(), 0, wrapped, encrypted.length, 4);
+
+    return wrapped;
+  }
+
+  /**
+   * Decrypts input data. The input composes of (msg, padding if needed, mac) and sequence num.
+   * The result is msg.
+   * @param data the input byte array
+   * @param offset the offset in input where the input starts
+   * @param len the input length
+   * @return the new decrypted byte array.
+   * @throws SaslException if error happens
+   */
+  public byte[] unwrap(byte[] data, int offset, int len) throws SaslException {
+    // get plaintext and seqNum
+    byte[] encrypted = new byte[len - 4];
+    byte[] peerSeqNum = new byte[4];
+    System.arraycopy(data, offset, encrypted, 0, encrypted.length);
+    System.arraycopy(data, offset + encrypted.length, peerSeqNum, 0, 4);
+    byte[] decrypted = decryptor.update(encrypted, 0, encrypted.length);
+
+    // get msg and mac
+    byte[] msg = new byte[decrypted.length - 10];
+    byte[] mac = new byte[10];
+    System.arraycopy(decrypted, 0, msg, 0, msg.length);
+    System.arraycopy(decrypted, msg.length, mac, 0, 10);
+
+    // modify msg length if padding
+    int msgLength = msg.length;
+    if (decryptor.getTransformation() == CipherTransformation.AES_CBC_NOPADDING) {
+      msgLength -= (int) msg[msgLength - 1];
+    }
+
+    // check mac integrity and msg sequence
+    if (!integrity.compareHMAC(mac, peerSeqNum, msg, 0, msgLength)) {
+      throw new SaslException("Unmatched MAC");
+    }
+    if (!integrity.comparePeerSeqNum(peerSeqNum)) {
+      throw new SaslException("Out of order sequencing of messages. Got: " + integrity.byteToInt
+          (peerSeqNum) + " Expected: " + integrity.peerSeqNum);
+    }
+
+    // return msg considering padding
+    if (msgLength == msg.length) {
+      return msg;
+    } else {
+      byte[] clearMsg = new byte[msgLength];
+      System.arraycopy(msg, 0, clearMsg, 0, msgLength);
+      return clearMsg;
+    }
+  }
+
+  private void checkTransformation(CipherTransformation transformation) throws IOException {
+    if (transformation == CipherTransformation.AES_CBC_NOPADDING
+        || transformation == CipherTransformation.AES_CTR_NOPADDING) {
+      return;
+    }
+    throw new IOException("AES cipher transformation is not supported: "
+        + transformation.getName());
+  }
+
+  /**
+   * Helper class for providing integrity protection.
+   */
+  private static class Integrity {
+
+    private int mySeqNum = 0;
+    private int peerSeqNum = 0;
+    private byte[] seqNum = new byte[4];
+
+    private byte[] myKey;
+    private byte[] peerKey;
+
+    Integrity(byte[] outKey, byte[] inKey) throws IOException {
+      myKey = outKey;
+      peerKey = inKey;
+    }
+
+    byte[] getHMAC(byte[] msg, int start, int len) throws SaslException {
+      seqNum = intToByte(mySeqNum ++);
```

Comment:
How about to reuse seqNum byte array without need a new one in intToByte?
As seqNum already initialized and it is a member.